### ISSUE TITLE
Cannot Select Element Using Picker and/or Enforce Cosmetic Filter

### ISSUE THREAD

**Issue by GuardianMajor** [uBlockOrigin/uBlock-issues#84]
### Prerequisites

- [x] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/)
    
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [x] uBlock Origin is the only extension
    - [x] uBlock Origin with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [x] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

Using the element picker, nothing happens (I have checked other reports and my cosmetic filtering is not disabled and never has been) and right clicking and choosing "Block element doesn't bring up the filter box and so I resorted to creating the filter manually and it won't enforce it, instead throws an error in the console: 

> this.f1 is undefined

pointing to: `static-net-filtering.js` Line:1211
```
FilterPair.prototype.match = function(url, tokenBeg) {
--> if ( this.f1.match(url, tokenBeg) === true ) {
        this.f = this.f1;
        return true;
    }
    if ( this.f2.match(url, tokenBeg) === true ) {
        this.f = this.f2;
        return true;
    }
    return false;
};
```

### A specific URL where the issue occurs

It has been happening in a lot of places, such as deviantart.com but for the sake of example, let's use the last link it happened on and I can provide more examples later if it becomes necessary to figure out why: http://therealstanlee.com/nine-scenes-from-superhero-films-that-were-supposed-to-happen-but-never-did/

### Steps to Reproduce

1. Go to the link
2. Click on the "dropper" and nothing happens
3. Right click and choose "Block element" and nothing happens
4. Go to "My filters"
5. Create a simple filter like `###quads-ad1` 
   (ideally something like `##[id^='quads-ad']` to get all of them in one filter)
(we are targeting this)
![image](https://user-images.githubusercontent.com/5274181/41803545-d5d06f64-763e-11e8-8f63-0ec9273a98c8.png)
6. Apply changes
7. Reload page and nothing happens
8. Check console

```
this.f1 is undefined (2) static-net-filtering.js:1211
this.f1 is undefined     static-net-filtering.js:1211
this.f1 is undefined (9) static-net-filtering.js:1211
this.f1 is undefined     static-net-filtering.js:1211
this.f1 is undefined     static-net-filtering.js:1211
```
![image](https://user-images.githubusercontent.com/5274181/41803603-6d09fe72-763f-11e8-9e2a-8b281d2eb1cc.png)

and for the record, this is what the panel looks like (showing what is blocked, which is the reason why the banger.js is breaking which is fine)
![image](https://user-images.githubusercontent.com/5274181/41803630-a962133c-763f-11e8-99e7-b151e0fb0f55.png)

### Expected behavior:

You'd expect to be able to use the dropper to pick and element and then build the filter. Alternatively you'd expect to be able to right click and block an element to bring up the filter box to build it. Finally you would expect that no matter what putting a manual filter would result in the items being blocked/scrubbed.

### Actual behavior:

None of the above expected behavior is happening.

### Your environment

* uBlock Origin version: v1.16.11rc0
* Browser Name and version: Firefox 56.0.2 (64-bit)
* Operating System and version: Windows 10 Pro (1803.b17134.112)

**Comment by GuardianMajor** [comment]
As promised, here is an example happening on deviantart.com: https://guardianmajor.deviantart.com/art/Atomic-Blonde-Have-a-Drink-712766717

(the grey box is what we are targeting)
![image](https://user-images.githubusercontent.com/5274181/41803844-418a7ba2-7642-11e8-80f8-e9ff7447178d.png)

simple filter `##.partial-ad,.dp-ad-target`

In the case of this page, no error in the console like before but you still can't pick element, right click block it or even get it to block using manual filter.

**Comment by uBlock-user** [comment]
I do not get any such errors on Chrome or Firefox Nightly. Are you on legacy version of Firefox ?

**Comment by BilBg** [comment]
This seems to be true for any page - e.g. this one here you are reading now.
I can't start picker or zapper (i.e. "nothing happens")
uBlock Origin v1.16.10
Chromium 50 - Windows XP
(I think in v1.16.8 this was OK, I may check if needed)

From background.html

```
Error in event handler for webRequest.onBeforeRequest/2: TypeError: Cannot read property 'match' of undefined
    at µBlock.staticNetFilteringEngine.FilterPair.match (chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/js/static-net-filtering.js:1211:17)
    at µBlock.staticNetFilteringEngine.FilterContainer.matchTokens (chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/js/static-net-filtering.js:2534:31)
    at µBlock.staticNetFilteringEngine.FilterContainer.matchString (chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/js/static-net-filtering.js:2770:31)
    at µBlock.PageStore.PageStore.filterRequest (chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/js/pagestore.js:640:46)
    at µBlock.webRequest.onBeforeRequest (chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/js/traffic.js:157:28)
    at onBeforeRequest (chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/js/vapi-webrequest.js:176:16)
```

**Comment by gorhill** [comment]
What are the filter lists selected + custom filters?

**Comment by gorhill** [comment]
Code reviewing the `FilterPair` code I definitely found a bug which could cause the propery `f1` to become `undefined`. I will assume this is the bug you are seeing and publish an emergency fix.

The key to reproduce this bug are the exact filter lists being enabled, including custom filters (rules are not important).

**Comment by uBlock-user** [comment]
is this bug specific to legacy branch ?

**Comment by gorhill** [comment]
Hmm correction, I do not see a bug. However from the messages reported in the console, the failure is in `FilterPair`. Again, key to reproduce is for me to known exactly the filter lists/custom filters.

**Comment by ana-ka** [comment]
Same issue for me, Firefox 60.0.2, ublock 1.16.10.

Filer lists used:

- All "built-in" **except** for ublock filters - Experimental
- All in Ads
- All in Privacy
- All in Malware domains **except** Spam404
- All in Annoyances
- Dan Pollock's hosts file and Peter Lowe's Ad and tacking server list
- EasyList Germany, Adguard Russian Filter and RU AdList

No custom filters or rules, except for probably unrelated Whitelist entries. Hope it helps, feel free to request more info.

**Comment by ana-ka** [comment]
@gorhill Basically any site. Just use this very Github-page or google.com. Clicking the small element-picker icon does nothing. Right-clicking something, choosing "block element" does nothing. All that used to work earlier, though I cannot tell what excactly broke it. I very rarely used the feature.

I will try to disable any other add-ons now and report back if that did anything.

**Comment by gorhill** [comment]
Code reviewing, the only way I can see one could end up with an invalid `FilterPair` data structure is that there is a `badfilter` which matches and removes both filters from `FilterPair`. The code assumes this can never occur (because duplicate are discarded), but apparently the bug report here seems to show this can happen. Again, reproducing on my side is key however.

**Comment by gorhill** [comment]
@ana-ka Try to force update all your filter lists, as I can't reproduce on my side I expect you won't be able either once you force update.

**Comment by ana-ka** [comment]
@gorhill Wow, good call. That indeed solved it. I am almost certain that I updated filter lists before though. What I did now was to purge caches and only then update. After that, everything is back to normal.

**Comment by gorhill** [comment]
Ok this confirms why I couldn't reproduce. Still, there is an issue lingering in there that I need to pinpoint.

**Comment by ana-ka** [comment]
I'll keep an eye on this issue. If there is anything I could try, let me know. Won't be here 24/7 though. Thanks so far! :)

**Comment by gwarser** [comment]
Original Poster is using outdated browsers.

**Comment by gorhill** [comment]
Still, the issue with `FilterPair` should not exist with FF56, I can't see why there would be an issue with it, the code just demand ES6-compliance, so FF56 should be fine.

**Comment by uBlock-user** [comment]
`v1.16.11rc0` is a WebExtension and OP is running that on FF56 which is the last legacy version, wouldn't that be the reason to begin with ?

**Comment by ana-ka** [comment]
One more thought - since I disabled several groups of filter lists, only leaving on the others, does this indicate that the filters triggering the issue were present in several filter lists across different groups? Maybe that's a starting point.

**Comment by gorhill** [comment]
@uBlock-user uBO is officially compatible with FF52+ and Chromium 45+. I need to find out what is causing the issues both @GuardianMajor and @BilBg see (though I won't be able to test with Chromium 50).

**Comment by jcartner-young** [comment]
Hi, guys.
First post, but I thought I should register and try since hardly anybody feedsback on the software they use and I have some programming knowledge.
I've been having this very same issue for a few weeks now.
Can we have some debug output or logs to send to you? - I can't get any runtime logging output from the uBlock plug-in.
I use uBlock to block the comments section on YouTube, but it's been hit and miss, with filters not various other options/interface elements not working/responding correctly.
Does Mozilla need to fix it's plug-in architecture?
I've had huge memory leaks from FF for years.
Could you add a few debugging options (what, where, how, etc) for us bug hunters so we can log our problems and you can see what's causing them??
So many questions.  Can anyone shed light on what is causing these errors?
- James

**Comment by UnicornDuckEli5** [comment]
For Firefox I have same issues mentioned above plus I can't update my filter list but everything works fine on Chrome

**Comment by gorhill** [comment]
@jcartner-young Submit a new issue please.

**Comment by gorhill** [comment]
@UnicornDuckEli5 Submit a new issue please.

Not all issues have to be presumed to be related top this one here, this will be determined once you submit your issue, with all the details necessary for other to reproduce.

**Comment by UnicornDuckEli5** [comment]
I didn't want to create duplicate because my issue is exact same with GuardianMajor and BilBg I just wanted to make an addition that when this issue occurred for me I purged my filter list and wasn't able to update after that then tried same for Chrome and it works fine I thought it might be related.

**Comment by gorhill** [comment]
> purged my filter list and wasn't able to update after that

What does the browser console report when you try to update? As asked above to others, what filter lists do you have enabled?

**Comment by UnicornDuckEli5** [comment]
No error in console, I have these enabled 
uBlock filters​​​​​
uBlock filters – Badware risks​
uBlock filters – Privacy
uBlock filters – Resource abuse​​​​​
uBlock filters – Unbreak
Adblock Warning Removal List​
Adguard Base Filters​
EasyList​
Adguard Spyware Filters​
EasyPrivacy​
Malware Domain List​​​​​​
Malware domains​
Peter Lowe’s Ad and tracking server list​

**Comment by uBlock-user** [comment]
I have all these filterlists + more and I can't reproduce on either of the browsers whatsoever..

**Comment by jcartner-young** [comment]
I am using Linux Debian x64, FF Quantum 60.0.2
It can work, but usually fails.

I haven't changed the enabled filter lists from default, just added a few custom content filters...

83,453 network filters ＋ 44,250 cosmetic filters from:
        My filters ​​​​​3 used out of 3​​​​ 
    Built-in (5/7)
        uBlock filters​ --- ​​​​6,486 used out of 6,541​​
​​        uBlock filters – Badware risks​ -- ​​​​14 used out of 14​​
​​        uBlock filters – Privacy​​​ -- ​​93 used out of 97​​
​​​​        uBlock filters – Resource abuse​​​​ -- ​259 used out of 261​​
​​​​        uBlock filters – Unbreak -- 445 used out of 469​​
Ads (1/4)
​​        EasyList​ -- ​​​​82,647 used out of 82,739​​
Privacy (1/3)
​​        EasyPrivacy​ -- ​​​​15,006 used out of 15,173​​
Malware domains (2/4)
​​        Malware Domain List​ -- ​​​​1,116 used out of 1,131​​
​​        Malware domains​ -- ​​​​23,079 used out of 23,079​​
Multipurpose (1/4)
​​        Peter Lowe’s Ad and tracking server list​ -- ​​​​2,636 used out of 2,636​​​​

Logger window remains empty after clicking on 'Block Element' in menu or dropper tool from toolbar window, just shows YouTube XHR requests firing.

Web client dev console remains empty after uBO purge & update.

uBlock Console after update reads...

17:22:35 |   |   |   | xhr | https://raw.githubusercontent.com/uBlockOrigin/uAssets/master/filters/resource-abuse.txt?_=212468
-- | -- | -- | -- | -- | --
17:22:34 |   |   |   | xhr | https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts&showintro=1&mimetype=plaintext&_=212468
17:22:34 |   |   |   | xhr | https://mirror.cedia.org.ec/malwaredomains/justdomains?_=212468
17:22:33 |   |   |   | xhr | https://cdn.rawgit.com/NanoMeow/MDLMirror/master/hosts.txt?_=212468
17:22:32 |   |   |   | xhr | https://easylist.to/easylist/easyprivacy.txt?_=212468
17:22:29 |   |   |   | xhr | https://easylist.to/easylist/easylist.txt?_=212468
17:22:28 |   |   |   | xhr | https://raw.githubusercontent.com/uBlockOrigin/uAssets/master/filters/unbreak.txt?_=212468
17:22:28 |   |   |   | xhr | https://raw.githubusercontent.com/uBlockOrigin/uAssets/master/filters/privacy.txt?_=212468
17:22:27 |   |   |   | xhr | https://raw.githubusercontent.com/uBlockOrigin/uAssets/master/filters/badware.txt?_=212468
17:22:26 |   |   |   | xhr | https://raw.githubusercontent.com/uBlockOrigin/uAssets/master/filters/filters.txt?_=212468
17:22:25 |   |   |   | xhr | https://raw.githubusercontent.com/gorhill/uBlock/master/assets/assets.json?_=212468

If uBlock /is/ reporting interface interactions, it's not seeing my menu item/window button clicks! :/

Can we enable deeper debug reporting?

**Comment by gorhill** [comment]
@jcartner-young Are you seeing `this.f1 is undefined` in the browser console?

If not, please open a new issue with your its details and repro steps.

**Comment by jcartner-young** [comment]
What the?!?!! It just worked - The preview pane popped up.
Still, no errors in console! O_o
I'll open a new issue then, though I'm not hopeful as there seems to be minimal output from the plug-in code.

**Comment by gorhill** [comment]
I've identified a way in which `this.f1 is undefined` can happen. The repro steps would involve creating custom static network filters using the element picker. Another condition is that the created custom static network filters are also a target of a matching `badfilter`. Now is to identify which specific custom filters can cause this in order to have solid repro steps.

**Comment by jcartner-young** [comment]
A few questions about this issue...
Shouldn't this.f1 have a default (detectable) safe state before doing anything with it?
What does the 'badfilter' define and why does it match?
What are the custom filters doing that's so badly affecting the interface?
Sounds like a very strange situation/race condition is arising, or the code is just overly complex :/
Is the plug-in creating bad rules from the picker/zapper?
I am extremely interested in fixing this plug-in as I install this on many systems and suggest this to everybody I meet.

**Comment by gorhill** [comment]
> I've identified a way in which

Well, I've tried to come up with synhtetic repro steps, but I can't figure any -- so in the end I may be wrong that there is a way to end up with an undefined `this.f1`. Back to square one: need repro steps from @GuardianMajor or @BilBg.

**Comment by gorhill** [comment]
I finally could come up with synthetic repro steps, but they are so contrived that I can't imagine how this could happen normally. Here:

- In "Filter lists" pane, un-select all filter lists, except for "My filters"
- In "My filters", add the filter `||ycombinator.com/favicon.ico$image`, click "Apply changes"
- Open `https://news.ycombinator.com/` in a new tab
- Open the logger
- Force a refresh of `https://news.ycombinator.com/` so that the logger is populated
- Click on any entry in the logger in order to access the "Static filtering" pane of any network request
- In the textarea of the "Static filtering" pane, paste the **two** following filters:
    - `||ycombinator.com/favicon.ico$image`
    - `||ycombinator.com/favicon.ico$image,badfilter`
- Click "Create"
- Force a refresh of `https://news.ycombinator.com/`

An exception occur at the line mentioned by OP, because `this.f1` is undefined.

**Comment by gorhill** [comment]
Following the exact steps is key, they are designed such that `FilterPair` ends up being used.

**Comment by uBlock-user** [comment]
I did just that and yet it doesn't throw that error on my end.

**Comment by gorhill** [comment]
You will have to sort this out on your side -- I did follow all the steps and got this in Chrome 67:

![a](https://user-images.githubusercontent.com/585534/41812393-0918449e-76f0-11e8-8a77-3ccabc2f14c0.png)

**Comment by gorhill** [comment]
> Can't sort it out if it doesn't happen

This is code, it's deterministic. There is no reason for it to not happen on your side, aside you having a different test environment. Start the steps with default uBO config -- and key to these is having no rules/custom filters before following every steps.

**Comment by uBlock-user** [comment]
I'm on Windows 10, that's the only difference. 

> Start the steps with default uBO config

Done and still the same, no errors/ no warnings, this one is proving difficult.

**Comment by gorhill** [comment]
Your screenshot shows you are looking at the page console. The issue is with uBO, not the page, you need to look at uBO's console.

**Comment by uBlock-user** [comment]
> you need to look at uBO's console.

That's missing from your STR, you should add that. Throws that error in the background.html console page.

**Comment by GuardianMajor** [comment]
@gorhill I am sorry for the delay coming back, I was having some health issues but it seems you are on top of it and have already found the issue, so unless I can provide something in the way of helping fix this that hasn't already been provided by everyone else, I will stay put.

Just a quick side note, this tends to sometimes snowball into completely killing all resources which means when you load a page, it is just garbled text, no CSS, no Images, no Scripts, nada. I thought I had a corrupted profile until I rolled it back to .10 and it fixed it, so I have a feeling something crept into the latest RC, cheers and thanks for looking into it.

**Comment by gwarser** [comment]
>when you load a page, it is just garbled text, no CSS, no Images, no Scripts, nada

Fanboy Annoyance list had issue with a wildcard

https://old.reddit.com/r/uBlockOrigin/comments/8sewas/fanboys_annoyance_list_blocking_most_websites/
https://old.reddit.com/r/uBlockOrigin/comments/8sf9nr/fanboyeasylistmerged_ultimate_list_is_blocking/

**Comment by gorhill** [comment]
>  I rolled it back to .10 and it fixed it

No [changes in 11rc0](https://github.com/gorhill/uBlock/compare/1.16.10...master) in the code to explain this, the biggest change is the one here, the other code change is a minor change in the popup panel. The only explanation is that of filter issue, which got solved in the process of reverting back.

What I was interested into is your custom filters, especially the `badfilter` ones, to find out and confirm my findings are related to the reported `this.f1` issue.

**Comment by BilBg** [comment]
> What are the filter lists selected + custom filters?

[my-ublock-backup_2018-06-24_02.51.21 - Problem ~.txt](https://github.com/uBlockOrigin/uBlock-issues/files/2130727/my-ublock-backup_2018-06-24_02.51.21.-.Problem.txt)

**Comment by BilBg** [comment]
> Try to force update all your filter lists, as I can't reproduce on my side with your lists, I expect you won't be able either once you force update

I have v1.16.10 disabled.  I can stop Internet, enable v1.16.10 and save cached lists in files (is there a faster way to do this? - I will need to click on every list and Copy/Paste the text)

Do you want/need this?

**Comment by gorhill** [comment]
> Do you want/need this?

No need, the issue can only manifest through the element picker or logger UI, by creating a static network filter, which is a duplicate of an already present filter in custom your filters (which was also maybe just created before the duplicate one), and which matches an existing `badfilter` filter somewhere, and to add even more to the unlikelihood, the created filter should be sitting alone in its bucket internally (i.e. extracted token from the filter is likely unique). At this point it's best to wait for the console to spit out the information if it ever occurs again.

**Comment by GuardianMajor** [comment]
@gwarser Thanks for the heads  up, but I don't use fanboy filters, I use the basic uBlock ones that are "in house" and the adguard ones and the "hosts" list ones, that's it and then whatever I make up on my own along the way to address more narrow cases. But it seems like Raymond figured out whatever it was, so we can hope it resolves it. Thanks again.

**Comment by BilBg** [comment]
To document how this (any) page was broken in the conditions of this issue 84
(uBO 1.16.10 + BilBg_Lists_issue84.zip)

 

![browser 6 - bad page](https://user-images.githubusercontent.com/22244214/42681966-dba3cea0-8691-11e8-93d5-c80a1b97f98e.png)

**Comment by BilBg** [comment]
Now in uBO 1.16.12 all i OK - and still using Old lists (you can find in `BilBg_Lists_issue84.zip`), i.e. this is unchecked:
- [ ] Auto-update filter lists

 
* Next is info from uBO 1.16.12

 
![cjpalhdlnbpafiamejdnhcphjbkeiagm-1531439735242-1](https://user-images.githubusercontent.com/22244214/42683387-e478598e-8695-11e8-81c3-88f9dcd732d7.png)

 
![cjpalhdlnbpafiamejdnhcphjbkeiagm-1531439735242-2](https://user-images.githubusercontent.com/22244214/42683450-0626a8f6-8696-11e8-8eb4-f219deb1cec1.png)

 
[cjpalhdlnbpafiamejdnhcphjbkeiagm-1531439735242.log](https://github.com/uBlockOrigin/uBlock-issues/files/2192050/cjpalhdlnbpafiamejdnhcphjbkeiagm-1531439735242.log)


`[12,"^https?:\\/\\/([0-9a-z\\-]+\\.)?(9anime|animeland|animenova|animeplus|animetoon|animewow|gamestorrent|goodanime|gogoanime|igg-games|kimcartoon|memecenter|readcomiconline|toonget|toonova|watchcartoononline)\\.[a-z]{2,4}\\/(?!([Ss]cripts|[Uu]ploads|[Ii]mages|combined|content|cover|img|static|thumbs|wp-content|wp-includes))(.*)"]`

**Comment by BilBg** [comment]
@gorhill 
     I don't know if the following is normal or a bug (so don't know if it needs new issue) - I see this "Errors" maybe for an year but don't "feel" any bad things to happen because of this.

Console of `background.html` is empty except for the following line which is normal for Windows XP:
`Failed to load resource: net::ERR_SSL_VERSION_OR_CIPHER_MISMATCH https://easylist.to/easylist/easyprivacy.txt?_=213222`

uBlock Origin v1.16.16
Chromium 50 - Windows XP

![ublocko - for node of nodes - 0](https://user-images.githubusercontent.com/22244214/44623062-2575b300-a8ce-11e8-873b-90c47f0eeb05.png)

![ublocko - for node of nodes - 1](https://user-images.githubusercontent.com/22244214/44623064-34f4fc00-a8ce-11e8-998d-725d2343028b.png)

![ublocko - for node of nodes - 2](https://user-images.githubusercontent.com/22244214/44623066-3d4d3700-a8ce-11e8-9946-06324e7ed008.png)

**Comment by gorhill** [comment]
Looks like Chrome 50 does not support `for...of` iterator for NodeList objects: <https://developer.mozilla.org/en-US/docs/Web/API/NodeList/entries#Browser_compatibility>.

This means I will have to bump up the minimum version to Chromium 51.

**Comment by BilBg** [comment]
Do you think that the Polyfills generated by "Polyfill service" can be used to restore compatibility?

https://cdn.polyfill.io/v2/polyfill.js
https://cdn.polyfill.io/v2/polyfill.min.js

https://cdn.polyfill.io/v2/polyfill.js?flags=gated
https://cdn.polyfill.io/v2/polyfill.min.js?flags=gated

https://cdn.polyfill.io/v2/polyfill.js?flags=gated&ua=chrome/50
https://cdn.polyfill.io/v2/polyfill.min.js?flags=gated&ua=chrome/50

https://cdn.polyfill.io/v2/polyfill.js?flags=gated&ua=chrome/45
https://cdn.polyfill.io/v2/polyfill.min.js?flags=gated&ua=chrome/45

The site:
https://polyfill.io/v2/docs/
https://github.com/Financial-Times/polyfill-service

**Comment by BilBg** [comment]
Here is an example how to Not load the Polyfills "For modern browsers" (I don't know enough to vote for correctness):
https://devhints.io/polyfill.io

Current version:
`<script>if(!(window.Promise&&[].includes&&Object.assign&&window.Map)){document.write('<script src="https://cdn.polyfill.io/v2/polyfill.min.js"></scr'+'ipt>')}</script>`

Previous version (was on the same page on 06.05.2018):
`<script>(function(d,s){
if(window.Promise&&[].includes&&Object.assign&&window.Map)return;
var sc=d.getElementsByTagName(s)[0],js=d.createElement(s);
js.src='https://cdn.polyfill.io/v2/polyfill.min.js';
sc.parentNode.insertBefore(js, sc);
}(document,'script'))</script>`

**Comment by jspenguin2017** [comment]
@BilBg Extensions cannot load remote code by default.

**Comment by jspenguin2017** [comment]
@nikrolls Just FYI in case you haven't abandoned the Edge port yet. `for ... of` loop on a static `NodeList` throws until Edge 40. Fixed in Edge 41.

**Comment by BilBg** [comment]
> @BilBg Extensions' privileged pages cannot load remote code by default.

No need to be "remote code", just save the file generated (`polyfill.js` or `polyfill.min.js`) and put in some of the extension directories and load from there (but I don't know if this will help the `for ... of` loop):
https://cdn.polyfill.io/v2/polyfill.js?flags=gated&ua=chrome/45
https://cdn.polyfill.io/v2/polyfill.min.js?flags=gated&ua=chrome/45

This is the extreme case - I think all the functions they have are "loaded" in the generated .js (`ua=chrome/7` and `ua=chrome/50` generates the same code/file when using flag `always`):
https://cdn.polyfill.io/v2/polyfill.js?features=default,es6,es5&flags=gated,always&ua=chrome/50
https://cdn.polyfill.io/v2/polyfill.min.js?features=default,es6,es5&flags=gated,always&ua=chrome/50

https://cdn.polyfill.io/v2/polyfill.js?features=default,es6,es5&flags=gated,always&ua=chrome/7
https://cdn.polyfill.io/v2/polyfill.min.js?features=default,es6,es5&flags=gated,always&ua=chrome/7

**Comment by jspenguin2017** [comment]
@BilBg I think it will work, I see some mention of `Symbol.iterator`. But XP support have to be removed one day, there is no reason that day can't be today. You can always fork, and with the polyfill, the maintenance work is basically zero for the near future.
